### PR TITLE
Document creation of peptides ml teaching data set 

### PR DESCRIPTION
**Pull request by taylorreiter**
I think all docs are in the md file for the PR. Feedback and additions welcome!

### CODE REVIEW

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+
+6. Convert the output TSV to a FASTA file that can be used to calculate peptide characteristics (molecular weight, etc).
+```
+cut -f2,3 peptides_for_ml_tmp.tsv | tail -n +2 | seqkit tab2fx > peptides_for_ml.fasta
+```
+
+7. Calculate peptide characteristics (length, mw, etc.)
+
+```
+curl https://raw.githubusercontent.com/Arcadia-Science/peptigate/main/scripts/characterize_peptides.py?token=GHSAT0AAAAAACPE43ZXXIARD4XPPPE2JMBWZPHWYPA
```
Comment (keithchev) [inline docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:149]: Probably don't want to include the token (even if its soon to expire)

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+peptide_characterization <- read_tsv("~/Downloads/backup_peptipedia_29_03_2023/peptides_for_ml_characterized.tsv")
+
+final_dataset <- final_dataset %>%
+  left_join(peptide_characterization, by = c("member" = "peptide_id")) %>%
+  mutate(length = nchar(member_sequence))
+
+write_tsv(final_dataset, "peptides_ml_dataset.tsv")
+```
+
+9. Check that the data set is predictive
```
Comment (keithchev) [inline docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:165]: would clarify to mention that this is checking that the dataset can be used to predict the bioactivity category.

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+Balanced Accuracy                  1.0000                               0.99803              0.84211        1.00000                        0.9919          1.00000                      0.99903            1.0000                0.9894                1.0
+```
+
+## Documentation of data set contents
+
+### Column descriptions
+
+* `rep`: representative peptide id for the cluster. Generated by running MMSeqs2 `linclust` on the peptipedia database with an identity threshold of 30%.
+* `member`: cluster member peptide id.
+* `member_sequence`: cluster member peptide sequence.
```
Comment (keithchev) [inline docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:233]: Could we use more descriptive names for these columns? Also, I think we could drop the `rep` column in the final dataset, right? (it might otherwise be confusing)

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+* `z1_lipophilicity`: Zscale quantifying the lipophilicity of the peptide.
+* `z2_steric_bulk_or_polarizability`: Zscale modeling steric properties like steric bulk and polarizability.
+* `z3_polarity_or_charge`: Zscale quantifying electronic properties like polarity and charge.
+* `z4_electronegativity_etc`: Zscale relating to electronegativity, heat of formation, electrophilicity and hardness.
+* `z5_electronegativity_etc`: Zscale relating to electronegativity, heat of formation, electrophilicity and hardness.
+* `length`: length in amino acids of the peptide.
+
+### Notes about the data set
+
+* We artificially generated similarity in peptides with specific bioactivity labels by clustering at 30% identity. The opposite of this should be done in real ML applications -- sequences that are similar in content should be removed so the same signal isn't repeated over and over again and so that there isn't pollution between the training and testing data sets.
```
Comment (keithchev) [inline docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:258]: I get this and it makes sense, but I think we should be careful not to bring this up in the lessons, because it's likely to be confusing to people.

**Review by keithchev — APPROVED** [review]
Looks good! This is very thorough (and definitely thorough enough for this context, imo).

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md (hunk 1 above)
Comment (keithchev) [reply docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:149]: edit: just remembered this repo is public 😬 so this token should be deleted asap.

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md (hunk 4 above)
Comment (taylorreiter) [reply docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:258]: kk i agree! i just wanted to add it so that _we_ know (and anyone who is a passerby on this repo) that this isn't a real dataset that is suddenly magic for peptide bioactivity prediction 😂

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md [hunk 5]
```diff
@@ -0,0 +1,264 @@
+## Documentation for how the dataset was made
+
+0. We used the environment defined in `make_peptides_ml_dataset.yml` to make the data set below.
+1. Download the peptipedia database:  https://drive.google.com/file/d/1x3yHNl8k5teHlBI2FMgl966o51s0T8i_/view?usp=sharing.
```
Comment (giladmishne) [inline docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:4]: Getting a 400 on that link:

image

Also, what is the original source of the data?

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+2. Unzip the database.
+3. The unzipped archive contains three files, `backup_sql.zip`, `peptipedia_csv.zip`, and `peptipedia_fasta.zip`.
+Unzipping `peptipedia_fasta.zip` results in a FASTA file `peptipedia.fasta` that contains all of the peptide sequences from the peptipedia database.
```
Comment (giladmishne) [inline docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:7]: OCD nit: indenting the content under the numbered bullets will improve the left justification to align with the bullets. I.e.

````
2. Unzip the database.
3. The unzipped archive contains three files, `backup_sql.zip`, `peptipedia_csv.zip`, and `peptipedia_fasta.zip`.
    Unzipping `peptipedia_fasta.zip` results in a FASTA file `peptipedia.fasta` that contains all of the peptide sequences from the peptipedia database.
    Unzipping `peptipedia_csv.zip` results in a CSV file `peptipedia.csv` that records the peptide id, peptide sequence, and predicted bioactivity labels.
    ...
````

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+Unzipping `peptipedia_csv.zip` results in a CSV file `peptipedia.csv` that records the peptide id, peptide sequence, and predicted bioactivity labels.
+Unzipping `backup_sql.zip` results in a postgres SQL dump file `backup_dump.sql`.
+This file contains metadata associated with the peptide sequences.
+To extract the metadata to separate CSV tables, we used the python code:
+```
```
Comment (giladmishne) [inline docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:12]: consider ```py for better rendering of the code

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+We will subsequently use this information to artificially generate a data set where the peptides or their characteristics are predictive of bioactivity labels.
+```
+mmseqs easy-cluster peptipedia_minlen10.fasta mmseqs/peptipedia_minlen10-0.3 tmp --min-seq-id 0.3
+```
+
+6. Filter to large clusters composed of peptides that all have the same bioactivity prediction and write a TSV file.
+Many of the peptides that end up in clusters have multiple bioactivity predictions (e.g. antimicrobial and therapeutic).
+The script below filters to peptide clusters that all have at least one bioactivity label that is the same.
+It also adds additional metadata (GO, PFAM) from the SQL database to the data.frame of peptide information.
+```
```
Comment (giladmishne) [inline docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:53]: consider ```r for better rendering of the code

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+# The goal of this script is to create a demo machine learning data set from peptides and their metadata.
+# It filters peptides in the peptipedia database to clusters of similar peptides
+# (30% identity or greater) that all have the same bioactivity label.
+# It then joins selected peptides to their metadata
+
+
+# filter to peptides clusters with many members that have the same bioactivity --------
+
+# read in cluster information from mmseqs2 linclust
+clusters <- read_tsv("~/Downloads/backup_peptipedia_29_03_2023/mmseqs/peptipedia_minlen10-0.3_cluster.tsv",
```
Comment (giladmishne) [inline docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:66]: is this the result of the previous step? (not explicitly mentioned, but maybe I'm missing it)

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+confusionMatrix(predictions, as.factor(testData$group_bioactivity))
+```
+
+which results in:
+```
+Confusion Matrix and Statistics
+
+                                Reference
+Prediction                       antimicrobial coagulation_or_anticoagulation growth_factor hormone immunological_activity metabolic neurological_activity propeptide signal_peptide therapeutic
+  antimicrobial                             67                              0             0       0                      0         0                     0          0              0           0
```
Comment (giladmishne) [inline docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:194]: this confusion matrix is a little "too good to be true", which either means that the task is "too easy" -- the classes are too easily-separable -- or that there's some unintended leak of the class into the features. for a dataset that is aimed at learning about ML, I wonder if it's better to have a harder task, one where a random forest with no tuning doesn't do better than 0.7-0.8 accuracy. that will allow showcasing the performance of different models, impact of hyperparams, interesting outliers, etc.

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+* `instability_index`: predicts the stability of a protein based on its dipeptide composition. A protein whose instability index is smaller than 40 is predicted as stable, a value above 40 predicts that the protein may be unstable.
+* `isoelectric_point`: the pH at which a particular molecule or surface carries no net electrical charge.
+* `molecular_weight`: the sum of the mass of each amino acid in the peptide sequence.
+* `pd1_residue_volume`: physical descriptor related to the peptide residue volume.
+* `pd2_hydrophilicity`: physical descriptor related to the peptide hydrophilicity.
+* `z1_lipophilicity`: Zscale quantifying the lipophilicity of the peptide.
+* `z2_steric_bulk_or_polarizability`: Zscale modeling steric properties like steric bulk and polarizability.
+* `z3_polarity_or_charge`: Zscale quantifying electronic properties like polarity and charge.
+* `z4_electronegativity_etc`: Zscale relating to electronegativity, heat of formation, electrophilicity and hardness.
+* `z5_electronegativity_etc`: Zscale relating to electronegativity, heat of formation, electrophilicity and hardness.
```
Comment (giladmishne) [inline docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:254]: what's the difference with the previous one?

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md (hunk 8 above)
Comment (taylorreiter) [reply docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md]: added

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md (hunk 10 above)
Comment (taylorreiter) [reply docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md]: ok I was able to decrease the out-of-the-box accuracy down to 92% (from 99% I believe) by decreasing the mmseqs cluster threshold from 30% to 10%, and by decreasing the minimum cluster size from 50 to 10. 

This is still not 70-80% accuracy. The things I can think to do to decrease the accuracy further are:
1. create separate training and test data sets explicitly, where sequences that cluster together won't be split between train and test. This should be pretty easy to do since we record cluster representatives, and will decrease the leakage.
2. I'm thresholding pretty stringently to only keep groups where all peptides have at least 1 bioactivity label that is the same for all peptides. I could drop that bar to something like 70% of peptides in a group, and just filter out the ones that don't have it or artificially assign the peptides in that group to have that label


Decreasing the group size to 10 also has the consequence of including less common groups. I think I potentially should also probably remove groups that are smaller than X (50? 200?) and decrease some of the homogeneity of the classes (cluster at 80% and select representatives?).

|group_bioactivity              |    n|
|:------------------------------|----:|
|signal_peptide                 | 3447|
|antimicrobial                  | 2254|
|therapeutic                    | 1352|
|immunological_activity         |  817|
|propeptide                     |  627|
|hormone                        |  421|
|growth_factor                  |  223|
|coagulation_or_anticoagulation |  222|
|neurological_activity          |  220|
|metabolic                      |   93|
|antibacterial_antibiotic       |   90|
|bioactive                      |   59|
|allergen                       |   56|
|toxins                         |   39|
|anti_gram                      |   33|
|amphibian_defense              |   29|
|anti_hiv                       |   27|
|cell_penetrating               |   26|
|defense                        |   26|
|cell_cell_communication        |   22|
|cell_sensing                   |   22|
|anticancer                     |   16|
|other_activity                 |   13|
|angiogenic                     |   12|
|antifungal                     |   12|
|nematode                       |   12|

I've never worked so hard to make a model less accurate 😂

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md (hunk 11 above)
Comment (taylorreiter) [reply docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:281]: idk :) I chose to report zscales in this script bc the other ones (1-3) report different information than other metrics calculated by the peptides package.

If you want more information, zscales come from this publication: https://pubs.acs.org/doi/10.1021/jm9700575
and zscales documentation in the peptides package is here: https://peptides.readthedocs.io/en/stable/api.html#peptides.ZScales

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md (hunk 6 above)
Comment (taylorreiter) [reply docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md]: oooh i love this and i didn't actually know this was a thing! thank you!

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md (hunk 9 above)
Comment (taylorreiter) [reply docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md]: yes! I added a comment and fixed the tool name to make this more clear.

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md (hunk 10 above)
Comment (taylorreiter) [reply docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md]: Ok I did two things:
1. Removed classes that only have one representative cluster --  all of these were sequences that all clustered together (10% or greater) in the initial clustering, meaning they are all similar. If I had included these, there wouldn't be any way to have data in the train and test sets that were from different clusters, so I just removed them.
2. split the test and training data sets so that sequences from the same clusters all ended up in the training or in the testing, so there wasn't pollution between the two data sets. this dramatically reduces the performance (as one would expect). Model performance summary is now:
   ```
   Overall Statistics
                                          
               Accuracy : 0.5808          
                 95% CI : (0.5566, 0.6047)
       No Information Rate : 0.2756          
       P-Value [Acc > NIR] : < 2.2e-16       
                                          
                     Kappa : 0.4825          
                                          
    Mcnemar's Test P-Value : NA   
   ```

not totally sure how to provide this data -- maybe as separate training and testing csv files?

Comment (taylorreiter) [reply docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md]: oh and here's the breakdown of the number of clusters (NOT number of sequences) that end up in each type, train and test.
```
                                 test train
  allergen                          1     1
  amphibian_defense                 1     1
  anti_hiv                          1     1
  antibacterial_antibiotic          2     3
  antimicrobial                    32    72
  bioactive                         1     2
  cell_penetrating                  1     1
  cell_sensing                      1     1
  coagulation_or_anticoagulation    2     2
  growth_factor                     3     5
  hormone                           7    14
  immunological_activity           12    25
  metabolic                         1     1
  neurological_activity             3     6
  propeptide                        6    14
  signal_peptide                   38    87
  therapeutic                      19    43
  toxins                            1     1
  ```

Comment (keithchev) [reply docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md]: This makes sense to me! FWIW, I would've cheated and just randomly shuffled a random subset of the rows in each of the feature columns. Obviously this would result in a patently "fake" dataset, but you'd be able to control the extent of the decreased accuracy by how many rows you shuffled. Anyway, I think this is a realistic and realistically hard dataset that will work for the AUG lessons. 

One issue: for some parts of the lessons it will be important to dynamically split the dataset into train and test (e.g. for doing cross-validation), so we can't really have static training and test sets. But I don't think splitting on cluster is necessary - the 58% accuracy feels a bit *too* low to me. As long as they're not huge, I would just commit the final merged dataset to the repo. If you want to retain the cluster-based train/test split, you could add a column containing a flag for this split, so we can recapitulate it in the lessons when we need an extra hard example.

Comment (taylorreiter) [reply docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md]: Hmmm ok. I could do the "right" thing that I would do if I were actually building a bioactivity classifier and cluster the entire data set at 80%. Then, I could still only select clusters that have consistent bioactivity labels (which would also reduce the size of the data set and increase signal although somewhat artificially), and then see how the classifier does. This will have the side effect of removing pollution between train and test no matter how it is carved up, and you won't need the representative sequence ID to make sense of things. This will be quick for me to try so I'll run it and report back!

Point well taken about shuffling rows -- would definitely be way more efficient. It makes me a little nauseous thinking of putting out fake data like that into the world tho...if I can pull off making this data set with semi-real data, then I'd like that, but if I don't accomplish the thing on this next iteration let's go with your approach 😂

Comment (taylorreiter) [reply docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md]: weeeeew it worked! I totally didn't think it was going to -- 

```
Overall Statistics
                                          
               Accuracy : 0.6188          
                 95% CI : (0.5805, 0.6559)
    No Information Rate : 0.2496          
    P-Value [Acc > NIR] : < 2.2e-16       
                                          
                  Kappa : 0.5316          
                                          
 Mcnemar's Test P-Value : NA   
 ```
 
 and here are the final classes that got included:
 ```
 |bioactivity              |   n|
|:------------------------|---:|
|antimicrobial            | 828|
|signal_peptide           | 783|
|propeptide               | 438|
|therapeutic              | 382|
|immunological_activity   | 307|
|hormone                  | 180|
|neurological_activity    | 111|
|antibacterial_antibiotic |  97|
|amphibian_defense        |  88|
|growth_factor            |  63|
|anti_gram                |  50|
```

Since we removed homogeneity, this data set should be able to be used any which way that anyone wants to. The only thing it can't be used for would be removing homogeneity, but even then you could just drop the threshold from 80% to 70% or 50% and it might still work

Comment (keithchev) [reply docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md]: Nice! This looks like it'll work. And it has the virtue of being entirely authentic (I agree generating a fake dataset doesn't *feel* great, and I hadn't thought about the fact that this will all be public, so the bar should be a bit higher than if we were just doing this for our own internal purposes).

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+               output_file = open(f'{output_file_prefix}{current_table}.csv', 'w')
+           elif line.startswith('\.'):
+               # End of data for the current table
+               if output_file:
+                   output_file.close()
+                   output_file = None
+           elif output_file:
+               # Convert tab-separated values to comma-separated values
+               csv_line = line.replace('\t', ',')
+               output_file.write(csv_line)
```
Comment (giladmishne) [inline docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:34]: can you note what files are expected to be extracted e.g. `table_public.gene_ontology` etc?

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+   
+   # The goal of this script is to create a demo machine learning data set from peptides and their metadata.
+   # It filters peptides in the peptipedia database to large clusters of similar peptides
+   # (80% identity or greater) that all have the same bioactivity label.
+   # It then joins selected peptides to their metadata.
+   
+   # filter to peptides clusters with many members that have the same bioactivity --------
+
+   # read in cluster information from mmseqs2 linclust
+   clusters <- read_tsv("~/Downloads/backup_peptipedia_29_03_2023/mmseqs/peptipedia_minlen10-0.8_cluster.tsv",
```
Comment (giladmishne) [inline docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:65]: should this be `10-0.1` from the previous step?

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+                        col_names = c("rep", "member"))
+   
+   # read in peptide bioactivity information
+   peptide_bioactivity <- read_csv('~/Downloads/backup_peptipedia_29_03_2023/peptipedia.csv') %>%
+     clean_names()
+   
+   # make the data set a little bit smaller by filtering to clusters that have > 5 members
```
Comment (giladmishne) [inline docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:72]: code says `> 2`

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+     pivot_longer(cols = starts_with("allergen"):last_col(), names_to = "bioactivity", values_to = "value")
+   
+   # Find clusters where all members have a 1 for any bioactivity
+   clusters_all_1_any_bioactivity <- peptide_bioactivity_with_clusters_long %>%
+     group_by(rep, bioactivity) %>%
+     summarise(all_ones = all(value == 1), .groups = 'drop') %>%
+     filter(all_ones) %>%
+     distinct(rep, .keep_all = TRUE)
+   
+   # filter to those clusters where all members have at least 1 bioactivity the same
```
Comment (giladmishne) [inline docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:104]: nit: "have at least 1 bioactivity the same" reads odd

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+9. Check that the data set is predictive of the label "group_bioactivity".
+   ```r
+   library(caret)
+   
+   set.seed(123)
+   
+   final_dataset <- read_tsv("peptides_ml_dataset.tsv")
+   final_dataset_filtered <- final_dataset %>%
+     select(-peptipedia_peptide_id, -peptide_sequence, -all_bioactivities, -go_id, -pfam_id,
+           -pfam_name, -pfam_type)
```
Comment (giladmishne) [inline docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:190]: why are these removed? should the pfam data be loaded in the first place?

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md [hunk 17]
```diff
@@ -0,0 +1,278 @@
...
+   
+   control <- trainControl(method="cv", number=10) # 10-fold cross-validation
+   model <- train(bioactivity ~ ., data=trainingData, method="rf", trControl=control)
+   
+   predictions <- predict(model, newdata=testData)
+   confusionMatrix(predictions, as.factor(testData$bioactivity))
+   ```
+
+   The `confusionMatrix()` command prints the following two results to the console.
+   The model has the following accuracy:
```
Comment (giladmishne) [inline docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:203]: maybe note that this accuracy is low by design? :)

**Review by giladmishne — APPROVED** [review]
looks great, thanks for working hard to make it perform worse :)

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md (hunk 10 above)
Comment (giladmishne) [reply docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md]: +1 this is great. thanks for the modifications!

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md (hunk 16 above)
Comment (taylorreiter) [reply docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:190]: PFAM and GO are removed because there are a lot of NAs, which random forests (caret) in R can't handle. I added them to the data/left them in there so we can teach about missing data and have more character data in the data set.

At the bottom of the MD file, I made some notes on how we could extend the data set to have more character data where we wouldn't have missing data -- BLAST, taxonomy, other annotation, etc. I haven't done that here bc of time/wanting to get a first product into everyone's hands so that we could start lesson dev, but we/I could still do it.

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md (hunk 14 above)
Comment (taylorreiter) [reply docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:72]: thank youuuu. I'll fix this. I was experimenting with different thresholds and 2 ended up working the best. Good catch!

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md (hunk 13 above)
Comment (taylorreiter) [reply docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:65]: wow i have all sorts of missed documentation updates, thank you! it's supposed to be 0.8 in the previous step, I'll fix that.

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md (hunk 12 above)
Comment (taylorreiter) [reply docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:34]: added! I'll also add the extracted CSVs to wherever I upload the database as well.

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md (hunk 15 above)
Comment (taylorreiter) [reply docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:104]: fixed!

File: docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md (hunk 17 above)
Comment (taylorreiter) [reply docs/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset.md:203]: added!

**Comment by taylorreiter** [comment]
Ok I addressed all of the feedback I received here (thank you both!) and made an OSF repo with the original files (https://osf.io/kb7z6/). We can still make changes to this data set, but I'm going to merge it for now!
